### PR TITLE
add init code reddit

### PR DESCRIPTION
**Pull request by hoang97**
- keywords list in file keywords.json
- result data in data.json
- infinite loop sleep for 1 second
- cache in online redis server

### CODE REVIEW

File: reddit/app.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from time import sleep
+from source.crawlers.reddit import RedditCrawler
+from source.utils.config import get_access, read, save, read_list
+import json
+import redis
+
+query_str = "crawl"
+data_path = "data.json"
+cache_db = redis.Redis(host='redis-10381.c135.eu-central-1-1.ec2.cloud.redislabs.com',
```
Comment (slaveeks) [inline reddit/app.py:9]: You need to create config for this data

File: reddit/app.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                       port=10381, db=0, password='EcqJqZRW8CPeSWaSvBpmN4vOpIh2tXoZ')
+
+try:
+    crawler = RedditCrawler(cache_db)
+    id = crawler.identify()
+    print(
```
Comment (slaveeks) [inline reddit/app.py:15]: Use logging for logs

**Comment by slaveeks** [comment]
remove ./reddit dir. and can you pull changes from main branch?

File: .gitignore [hunk 3]
```diff
@@ -5,3 +5,5 @@ redis_data
 .DS_Store
 venv
 config.json
+
+config.json
```
Comment (slaveeks) [inline .gitignore:9]: You duplicate `config.json` in .gitignore

File: crawlers/reddit/config.json.template [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+    "data_path": "./data.json",
+    "keywords_path": "./keywords.json",
```
Comment (slaveeks) [inline crawlers/reddit/config.json.template:3]: You already have keywords in this config file, on the 11 line

File: crawlers/reddit/main.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+while True:
+    keywords = read_list()
+    result = {}
+    for keyword in keywords:
+        result[keyword] = crawler.hot_posts(keyword)
+
+    save_list(result)
+
+    sleep(1)
```
Comment (slaveeks) [inline crawlers/reddit/main.py:24]: Get this value from config

File: crawlers/reddit/source/crawler.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+import requests.auth
+import json
+from source.database.cache import CacheDB
+from source.utils.tools import read_access_token, get_access_token, update_config, read_config
+from source.utils.log import logger
+
+
+class RedditCrawler:
+
+    def __init__(self, cache: CacheDB) -> None:
```
Comment (slaveeks) [inline crawlers/reddit/source/crawler.py:11]: Remove return type `None`

File: crawlers/reddit/source/crawler.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        self.cache = cache
+        self.access_token = read_access_token()
+        self.base_url = "https://oauth.reddit.com/"
```
Comment (slaveeks) [inline crawlers/reddit/source/crawler.py:14]: Let's move this data to const on the top of the file, like `BASE_URL`

File: crawlers/reddit/source/crawler.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            return True
+
+        return False
+
+    def _update_access_token(self) -> None:
+
+        access_token = get_access_token(
+            self.conf["USER_NAME"], self.conf["PASSWORD"], self.conf["CLIENT_ID"], self.conf["CLIENT_SECRET"])
+
+        update_config(access_token=access_token)
```
Comment (slaveeks) [inline crawlers/reddit/source/crawler.py:37]: I think, that it is not good idea to change config file from code

Comment (slaveeks) [reply crawlers/reddit/source/crawler.py:37]: We can also update `self.access_token`

File: crawlers/reddit/source/crawler.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            response = requests.get(
+                url, params=params, headers=self.headers).json()['names']
+            self._save_to_cache(url, params, response)
+
+        return response
+
+    def hot_posts(self, subreddit: str, g: str = "GLOBAL", after: str = None, before: str = None, count: int = 0, limit: int = 25, show: str = None, sr_detail: str = None) -> list:
+        '''List hotest posts from specific subreddit
+
+        :param `g`: one of (GLOBAL, US, AR, AU, BG, CA, CL, CO, HR, CZ, FI, FR, DE, GR, HU, IS, IN, IE, IT, JP, MY, MX, NZ, PH, PL, PT, PR, RO, RS, SG, ES, SE, TW, TH, TR, GB, US_WA, US_DE, US_DC, US_WI, US_WV, US_HI, US_FL, US_WY, US_NH, US_NJ, US_NM, US_TX, US_LA, US_NC, US_ND, US_NE, US_TN, US_NY, US_PA, US_CA, US_NV, US_VA, US_CO, US_AK, US_AL, US_AR, US_VT, US_IL, US_GA, US_IN, US_IA, US_OK, US_AZ, US_ID, US_CT, US_ME, US_MD, US_MA, US_OH, US_UT, US_MO, US_MN, US_MI, US_RI, US_KS, US_MT, US_MS, US_SC, US_KY, US_OR, US_SD)
```
Comment (slaveeks) [inline crawlers/reddit/source/crawler.py:116]: I think, we can omit this `GLOBAL, US, AR, AU, BG, CA, CL, CO, HR, CZ, FI, FR, DE, GR, HU, IS, IN, IE, IT, JP, MY, MX, NZ, PH, PL, PT, PR, RO, RS, SG, ES, SE, TW, TH, TR, GB, US_WA, US_DE, US_DC, US_WI, US_WV, US_HI, US_FL, US_WY, US_NH, US_NJ, US_NM, US_TX, US_LA, US_NC, US_ND, US_NE, US_TN, US_NY, US_PA, US_CA, US_NV, US_VA, US_CO, US_AK, US_AL, US_AR, US_VT, US_IL, US_GA, US_IN, US_IA, US_OK, US_AZ, US_ID, US_CT, US_ME, US_MD, US_MA, US_OH, US_UT, US_MO, US_MN, US_MI, US_RI, US_KS, US_MT, US_MS, US_SC, US_KY, US_OR, US_SD`

File: crawlers/reddit/source/utils/tools.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    for key in kwargs.keys():
+        if key in conf.keys():
+            conf[key] = kwargs[key]
+
+    with open(config_path, "w") as f:
+        json.dump(conf, f)
+
+
+def get_access_token(username, password, client_id, client_secret):
```
Comment (slaveeks) [inline crawlers/reddit/source/utils/tools.py:45]: What do you think about creating a separate class for login and working with access token?

File: crawlers/reddit/config.json.template [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    "token_path": "./access_token.json",
```
Comment (slaveeks) [inline crawlers/reddit/config.json.template:4]: Remove this

File: crawlers/reddit/config.json.template [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    "CLIENT_ID": "",
```
Comment (slaveeks) [inline crawlers/reddit/config.json.template:5]: Why do you use the uppercase here?

File: crawlers/reddit/main.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+from time import sleep
+from source.database.cache import CacheDB
+from source import RedditCrawler
+from source.utils.tools import read_list, read_config, save_list
+from source.utils.log import logger
+
+redis_creds = read_config().get('redis_creds')
```
Comment (slaveeks) [inline crawlers/reddit/main.py:7]: Why it is not in main function?

File: crawlers/reddit/source/database/cache.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+import redis
+from source.utils.log import logger
+
+
+class CacheDB:
+    '''Methods to work with Redis cache database'''
+
+    def __init__(self, host: str, port: str, password: str = '') -> None:
```
Comment (slaveeks) [inline crawlers/reddit/source/database/cache.py:8]: Remove return type `None`

File: crawlers/reddit/source/utils/tools.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+config_path = (Path(__file__).absolute().parent.parent.parent/'config.json')
+data_path = (Path(__file__).absolute().parent.parent.parent/'data.json')
+
+def read_config():
+    with open(config_path, 'r') as f:
+        data = json.load(f)
+        return data
+
+def save_list(data):
+    with open(data_path, "w") as f:
```
Comment (slaveeks) [inline crawlers/reddit/source/utils/tools.py:16]: Use single quotes

File: crawlers/reddit/source/utils/tools.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def read_list():
+    with open(config_path, 'r') as f:
+        data = json.load(f)
+        return data.get('keywords')
+
+
+def read_access_token():
+    with open(config_path, "r") as f:
```
Comment (slaveeks) [inline crawlers/reddit/source/utils/tools.py:27]: Use single quotes

File: crawlers/reddit/source/utils/tools.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+def update_config(**kwargs):
+    '''Update config file
+    params must be specify'''
+    conf = read_config()
+
+    for key in kwargs.keys():
+        if key in conf.keys():
+            conf[key] = kwargs[key]
+
+    with open(config_path, "w") as f:
```
Comment (slaveeks) [inline crawlers/reddit/source/utils/tools.py:41]: Use single quotes

File: crawlers/reddit/source/utils/tools.py [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        json.dump(conf, f)
+
+
+def get_access_token(username, password, client_id, client_secret):
+    url = "https://www.reddit.com/api/v1/access_token"
+    client_auth = requests.auth.HTTPBasicAuth(client_id, client_secret)
+    post_data = {
+        "grant_type": "password",
```
Comment (slaveeks) [inline crawlers/reddit/source/utils/tools.py:49]: Can you use only single quotes in your code?

File: crawlers/reddit/data.json [hunk 19]
```diff
@@ -0,0 +1,193 @@
+{
```
Comment (slaveeks) [inline crawlers/reddit/data.json:1]: Remove data)

File: crawlers/reddit/source/database/cache.py (hunk 14 above)
Comment (slaveeks) [reply crawlers/reddit/source/database/cache.py:8]: I did not remove it)

Comment (hoang97) [reply crawlers/reddit/source/database/cache.py:8]: oh sorry I forgot this file

File: crawlers/reddit/data.json (hunk 19 above)
Comment (slaveeks) [reply crawlers/reddit/data.json:1]: We have no need in it

**Review by slaveeks — APPROVED** [review]

**Review by n0str — APPROVED** [review]
Approve, but see some comments. You can fix them within this request or open a new one

**Pull request merged**